File: CSharp/Applications/AvaloniaLoggingDI/ViewLocator.cs
using Avalonia.Controls;
using Avalonia.Controls.Templates;
using AvaloniaLoggingDI.ViewModels;
using System;

namespace AvaloniaLoggingDI;

public class ViewLocator : IDataTemplate
{
    public IControl Build(object data)
    {
        string name = data.GetType().FullName!.Replace("ViewModel", "View");
        Type? type = Type.GetType(name);

        if (type != null)
        {
            return (Control)Activator.CreateInstance(type)!;
        }

        return new TextBlock { Text = "Not Found: " + name };
    }

    public bool Match(object data)
    {
        return data is ViewModelBase;
    }
}

File: VB/Applications/AvaloniaSerilogVBDI/ViewLocator.vb
Imports Avalonia.Controls
Imports Avalonia.Controls.Templates

Public Class ViewLocator : Implements IDataTemplate

    Public Function Build(data As Object) As IControl Implements ITemplate(Of Object, IControl).Build

        Dim name As String = data.GetType().FullName.Replace("ViewModel", "View")
        Dim type As Type = Type.GetType(name)

        If type IsNot Nothing Then
            Return DirectCast(Activator.CreateInstance(type), Control)
        End If

        Return New TextBox With {.Text = "Not Found: " + name}

    End Function

    Public Function Match(data As Object) As Boolean Implements IDataTemplate.Match

        Return TypeOf data Is ViewModelBase

    End Function

End Class